documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md

```diff
@@ -0,0 +1,586 @@
+- `indexer.py` will contain the logic for the actual indexing task. It will be
+  refactored from the existing
+  [`indexer.py`](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/ingestion_server/ingestion_server/indexer.py);
+  specifically all we need is the
+  [`replicate`](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/ingestion_server/ingestion_server/indexer.py#L157)
+  function.
+- `elasticsearch_models.py`, pulled from the file of the
+  [same name](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/ingestion_server/ingestion_server/elasticsearch_models.py)
+  in the ingestion server. Defines a mapping from a database record to an
+  Elasticsearch document.
+- Utility files for helper functions for connecting to Elasticsearch and
+  Postgres (e.g.
+  [`es_helpers.py`](https://github.com/WordPress/openverse/blob/main/ingestion_server/ingestion_server/es_helpers.py))
+
+The
+[`Dockerfile`](https://github.com/WordPress/openverse/blob/main/ingestion_server/Dockerfile)
+can be copied from the existing ingestion server. It should be updated to
+reference the new file structure, and to expose only a single port, which should
+be distinguished from the ports currently in use by the ingestion server (8001
+and 8002). Other necessary files, including `env.docker`, `.dockerignore`,
+`Pipfile`, and `gunicorn.conf.py` can all be copied in from the existing
+ingestion server as well.
+
+Finally we will update the monorepo's root
+[`docker-compose.yml`](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/docker-compose.yml)
+to add a new `catalog-indexer-worker` service. Its build context should point to
+the nested `data_refresh/indexer_worker` directory, and it should map the
+exposed port to enable the API to be reached by the catalog.
+
+When this work is complete, it should be possible to run `just catalog/ipython`
+and curl the new indexer worker. The existing ingestion-server and
+indexer-worker services are unaffected (it is still possible to run legacy data
+refreshes locally and in production).
+
+### Implement distributed reindexing locally
+
+In this step we will add tasks to the data refresh DAGs to orchestrate the
+distributed reindex. At the end of this step, it will be possible to run a
+distributed reindex locally. Because the infrastructure work to create the EC2
+instances is not complete, it can not be run on production yet. The following
+code can all be refactored from
+[`distributed_reindex_scheduler.py`](https://github.com/WordPress/openverse/blob/main/ingestion_server/ingestion_server/distributed_reindex_scheduler.py).
+
+- Use Airflow's
+  [EC2Hook#describe_instances](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/hooks/ec2/index.html#airflow.providers.amazon.aws.hooks.ec2.EC2Hook.describe_instances)
```

Comment:
This is a total bikeshed-y nit and feel free to ignore, but I prefer to use the [scope resolution operator](https://learn.microsoft.com/en-us/cpp/cpp/scope-resolution-operator?view=msvc-170) from C++ to denote member methods in classes (rather than `#` since that has a different meaning in Python). We also use it already in a few places: [1](https://github.com/WordPress/openverse/blob/ed267d80e8a67ef31d82f3857dce239fba45ae57/catalog/dags/common/sensors/utils.py#L36-L37), [2](https://github.com/WordPress/openverse/blob/ed267d80e8a67ef31d82f3857dce239fba45ae57/catalog/dags/data_refresh/create_filtered_index_dag.py#L99), [3](https://github.com/WordPress/openverse/blob/ed267d80e8a67ef31d82f3857dce239fba45ae57/catalog/dags/data_refresh/create_filtered_index_dag.py#L109). 

```suggestion
  [EC2Hook::describe_instances](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/hooks/ec2/index.html#airflow.providers.amazon.aws.hooks.ec2.EC2Hook.describe_instances)
```